sync-core/src/main/java/com/cloudant/sync/datastore/AttachmentManager.java

```diff
@@ -348,47 +382,172 @@ protected void purgeAttachments(SQLDatabase db) {
+        String filename = null;
+
+        db.beginTransaction();
+        try {
+            Cursor c = db.rawQuery(SQL_FILENAME_LOOKUP_QUERY, new String[]{ keyString });
+            if (c.moveToFirst()) {
+                filename = c.getString(0);
+                logger.finest(String.format("Found filename %s for key %s", filename, keyString));
+            } else if (allowCreateName) {
+                filename = generateFilenameForKey(db, keyString);
+                logger.finest(String.format("Added filename %s for key %s", filename, keyString));
+            }
+            c.close();
+            db.setTransactionSuccessful();
+        } catch (SQLException e) {
+            logger.log(Level.WARNING, "Couldn't read key,filename mapping database", e);
+            filename = null;
+        } finally {
+            db.endTransaction();
+        }
+
+        if (filename != null) {
+            return new File(attachmentsDir, filename);
+        } else {
+            // generateFilenameForKey throws an exception if we couldn't generate, this
+            // means we couldn't get one from the database.
+            throw new AttachmentException("Couldn't retrieve filename for attachment");
+        }
+    }
+
+
+    /**
+     * Iterate candidate filenames generated from the filenameRandom generator
+     * until we find one which doesn't already exist.
+     *
+     * We try inserting the new record into attachments_key_filename to find a
+     * unique filename rather than checking on disk filenames. This is because we
+     * can make use of the fact that this method is called on a serial database
+     * queue to make sure our name is unique, whereas we don't have that guarantee
+     * for on-disk filenames. This works because filename is declared
+     * UNIQUE in the attachments_key_filename table.
+     *
+     * We allow up to 200 random name generations, which should give us many millions
+     * of files before a name fails to be generated and makes sure this method doens't
+     * loop forever.
+     *
+     * @param db database to use
+     * @param keyString blob's key
+     */
+    static String generateFilenameForKey(SQLDatabase db, String keyString)
+            throws NameGenerationException {
+
+        String filename = null;
+
+        long result = -1;  // -1 is error for insert call
+        int tries = 0;
+        while (result == -1 && tries < 200) {
+            byte[] randomBytes = new byte[20];
+            filenameRandom.nextBytes(randomBytes);
+            String candidate = keyToString(randomBytes);
+
+            ContentValues contentValues = new ContentValues();
+            contentValues.put("key", keyString);
+            contentValues.put("filename", candidate);
+            result = db.insert(ATTACHMENTS_KEY_FILENAME, contentValues);
+
+            if (result != -1) {  // i.e., insert worked, filename unique
+                filename = candidate;
+            }
+
+            tries++;
```

Comment:
I tend to prefer it within the loop body rather than condition, but it's good to know. I guess I use `++` in a `for` loop's condition all the time so I don't really have a justification for my preference :)